Bug Report: MySQL 8.4 default configuration contains removed configuration variable `default_authentication_plugin`

### Overview of the Issue

The file `mysql84.cnf` [contains](https://github.com/vitessio/vitess/blob/b1620481df197d50110ce37c70e9a7c51b5fd0cc/config/mycnf/mysql84.cnf#L17) a reference to the configuration variable `default_authentication_plugin`. However, this variable has been removed starting in MySQL 8.4 as mentioned in the documentation here:

> The default_authentication_plugin system variable, deprecated in MySQL 8.0.27, is removed as of MySQL 8.4.0. Use authentication_policy instead.
>
> -- https://dev.mysql.com/doc/refman/8.4/en/mysql-nutshell.html

When using `mysqlctld` to start a mysqld process with version 8.4, the process fails due to this
reason. The following line is printed to the `error.log` file inside the respective mysqld's
directory: (e.g. `/vt/vt_0000000100/error.log`)

```
2024-07-18T03:18:42.364068Z 0 [ERROR] [MY-000067] [Server] unknown variable 'default_authentication_plugin=mysql_native_password'.
```


### Reproduction Steps

1. Build a Docker image using the following Dockerfile. Or run the commands inside the Dockerfile on a Rocky Linux 9 machine
	```
	$ cat Dockerfile
	FROM rockylinux:9

	# Install MySQL yum repository
	WORKDIR /tmp
	RUN dnf update -y
	RUN dnf install -y wget which make git
	RUN wget https://dev.mysql.com/get/mysql84-community-release-el9-1.noarch.rpm
	RUN dnf install -y mysql84-community-release-el9-1.noarch.rpm
	RUN dnf update -y
	RUN dnf install -y mysql-community-server

	# Install and add Go to the path
	RUN wget https://go.dev/dl/go1.22.5.linux-amd64.tar.gz
	RUN rm -rf /usr/local/go && tar -C /usr/local -xzf go1.22.5.linux-amd64.tar.gz
	RUN touch /root/.bashrc
	RUN echo 'PATH="$PATH:/usr/local/go/bin"' >> /root/.bashrc

	RUN wget https://github.com/etcd-io/etcd/releases/download/v3.5.14/etcd-v3.5.14-linux-amd64.tar.gz
	RUN tar zxf etcd-v3.5.14-linux-amd64.tar.gz
	RUN cp etcd-v3.5.14-linux-amd64/{etcd,etcdutl,etcdctl} /usr/local/bin

	# Switch to non-root user
	# From: https://github.com/vitessio/vitess/blob/c9c0aa5db40eaf508c5d124898c2af25a8b78cc2/docker/lite/Dockerfile#L22-L26
	RUN groupadd -r vitess && useradd -r -g vitess vitess
	RUN mkdir -p /vt/vtdataroot /home/vitess
	RUN chown -R vitess:vitess /vt /home/vitess

	RUN groupadd -r admin
	RUN usermod -a -G admin root
	RUN usermod -a -G admin vitess
	RUN chown vitess:admin /usr/local/bin

	RUN cp /root/.bashrc /home/vitess/.bashrc
	RUN chown -R vitess:vitess /home/vitess/.bashrc

	USER vitess
	RUN git config --global --add safe.directory /src/vitess%

	$ docker build -t go-and-mysqld-8-4:latest .
	```

2. Run a Docker container using the above image. Mount the Vitess source code as a volume
	```
	$ docker run --rm -it -v $(pwd):/src/vitess -w /src/vitess --name vitess-config-update go-and-mysqld-8-4 /bin/bash
	```
3. Build Vitess inside the Docker container
	```
	make clean; make install NOVTADMINBUILD=true PREFIX=/usr/local;
	mysqlctld --version;
	```
4. Run the basic local cluster setup script
	```
	cd examples/local/;
	VTDATAROOT=/vt bash 101_initial_cluster.sh
	```
5. Observe the following log regarding `mysqld` being unable to start in the output
	```
	F0718 03:41:12.441304   11775 mysqlctl.go:27] failed init mysql: /usr/sbin/mysqld: exit status 1, output:
	```
6. Observe that the following log lines are printed inside the file `/vt/vt_0000000101/error.log`
	```
	2024-07-18T03:41:10.841756Z 0 [ERROR] [MY-000067] [Server] unknown variable 'default_authentication_plugin=mysql_native_password'.
	```

(Due to the structure of [this code](https://github.com/vitessio/vitess/blob/b1620481df197d50110ce37c70e9a7c51b5fd0cc/go/vt/mysqlctl/mysqld.go#L932-L966), it is not possible to _remove_ the base configuration in this repository, which is embedded into the binary. So, I believe that this change should be back ported to v20, which is the first version which has a separate configuration for MySQL 8.4. Please correct me if there is another workaround for this which can be used with the published binaries of Vitess.)


### Binary Version

```sh
[vitess@0997b3d45ff7 local]$ mysqlctld --version
mysqlctld version Version: 21.0.0-SNAPSHOT (Git revision b1620481df197d50110ce37c70e9a7c51b5fd0cc branch 'main') built on Thu Jul 18 03:38:26 UTC 2024 by vitess@0997b3d45ff7 using go1.22.5 linux/amd64
```

Additionaly, the version of MySQL is also relevant:

```
[vitess@78ec5221d252 vitess]$ mysqld --version
/usr/sbin/mysqld  Ver 8.4.1 for Linux on x86_64 (MySQL Community Server - GPL)
```


### Operating System and Environment details

```sh
[vitess@0997b3d45ff7 local]$ cat /etc/os-release
NAME="Rocky Linux"
VERSION="9.4 (Blue Onyx)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="9.4"
PLATFORM_ID="platform:el9"
PRETTY_NAME="Rocky Linux 9.4 (Blue Onyx)"
ANSI_COLOR="0;32"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:rocky:rocky:9::baseos"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
SUPPORT_END="2032-05-31"
ROCKY_SUPPORT_PRODUCT="Rocky-Linux-9"
ROCKY_SUPPORT_PRODUCT_VERSION="9.4"
REDHAT_SUPPORT_PRODUCT="Rocky Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="9.4"
[vitess@0997b3d45ff7 local]$ uname -sr
Linux 6.6.16-linuxkit
[vitess@0997b3d45ff7 local]$ uname -m
x86_64
```


### Log Fragments

```
This is the complete `error.log` file from the `mysql` daemon process:


[vitess@0997b3d45ff7 vitess]$ cat /vt/vt_0000000100/error.log
2024-07-18T03:41:08.636300Z 0 [System] [MY-015017] [Server] MySQL Server Initialization - start.
2024-07-18T03:41:08.647320Z 0 [Warning] [MY-011068] [Server] The syntax 'skip_slave_start' is deprecated and will be removed in a future release. Please use skip_replica_start instead.
2024-07-18T03:41:08.648878Z 0 [System] [MY-013169] [Server] /usr/sbin/mysqld (mysqld 8.4.1) initializing of server in progress as process 11823
2024-07-18T03:41:08.689247Z 1 [System] [MY-013576] [InnoDB] InnoDB initialization has started.
get_mempolicy: Function not implemented
2024-07-18T03:41:08.700467Z 1 [Warning] [MY-011873] [InnoDB] Failed to set NUMA memory policy to MPOL_INTERLEAVE: Function not implemented
get_mempolicy: Function not implemented
2024-07-18T03:41:08.702826Z 0 [Warning] [MY-011879] [InnoDB] Failed to set NUMA memory policy of buffer pool page frames with mbind(0x7fffe7d1f000,137236480,MPOL_INTERLEAVE,...,...,MPOL_MF_MOVE) failed with Function not implemented
2024-07-18T03:41:08.706529Z 1 [Warning] [MY-011875] [InnoDB] Failed to set NUMA memory policy to MPOL_DEFAULT: Function not implemented
2024-07-18T03:41:09.025849Z 1 [System] [MY-013577] [InnoDB] InnoDB initialization has ended.
2024-07-18T03:41:09.429592Z 0 [Warning] [MY-013501] [Server] Ignoring --plugin-load[_add] list as the server is running with --initialize(-insecure).
2024-07-18T03:41:10.841756Z 0 [ERROR] [MY-000067] [Server] unknown variable 'default_authentication_plugin=mysql_native_password'.
2024-07-18T03:41:10.842514Z 0 [ERROR] [MY-013236] [Server] The designated data directory /vt/vt_0000000100/data/ is unusable. You can remove all files that the server added to it.
2024-07-18T03:41:10.842528Z 0 [ERROR] [MY-010119] [Server] Aborting
2024-07-18T03:41:12.384742Z 0 [System] [MY-015018] [Server] MySQL Server Initialization - end.
```

--- a/config/mycnf/mysql84.cnf
+++ b/config/mycnf/mysql84.cnf
@@ -14,7 +14,7 @@ mysqlx = 0
 # 8.4 changes the default auth-plugin to caching_sha2_password and
 # disables mysql_native_password by default.
 mysql_native_password = ON
-default_authentication_plugin = mysql_native_password
+authentication_policy = 'mysql_native_password'
 
 # Semi-sync replication is required for automated unplanned failover
 # (when the primary goes away). Here we just load the plugin so it's

